src/utilities/structured-output/structured-output.factory.ts

```diff
@@ -0,0 +1,169 @@
+import { OASResult, OperationExampleResult } from '../../results';
+import { StructuredOutput } from './structured-output.interface';
+
+class StructuredOutputFactory {
+  static buildFromOASResult(input: OASResult): StructuredOutput {
+    const testResults = input.results;
+
+    // assemble authenticationType string
+    const authenticationType = input.securitySchemes
+      .map((x) => x.type)
+      .join('/');
+
+    // assemble top-level output
+    const output: StructuredOutput = {
+      id: input.testName,
+      config: {
+        oasPath: String(input.oasPath),
+        server: String(input.server),
+        authenticationType: String(authenticationType),
+      },
+      error: input.error,
+      results: undefined,
+    };
+
+    if (!input.error && testResults) {
+      // assemble api-level output
+      this.assembleApiLevel(testResults, output);
+    }
+
+    return output;
+  }
+
+  private static assembleApiLevel(
+    testResults: OperationExampleResult[],
+    output: StructuredOutput,
+  ): void {
+    // calculate pass/warn/fail totals
+    const passCount = testResults.filter(
+      (result) => result.failures.size === 0,
+    ).length;
+    const warnCount = testResults.filter(
+      (result) => result.warnings.size > 0,
+    ).length;
+    const failCount = testResults.filter(
+      (result) => result.failures.size > 0,
+    ).length;
+    const totalCount = testResults.length;
+
+    // assemble api summary
+    output.results = {
+      apiSummary: {
+        totalPass: Number(passCount),
+        totalWarn: Number(warnCount),
+        totalFail: Number(failCount),
+        totalRun: Number(totalCount),
+        runDateTime: new Date(),
+      },
+    };
+
+    const operationIds = [
+      ...new Set(testResults.map((result) => result.operationId)),
+    ];
+    operationIds.forEach((operationId) => {
+      // iterate through all distinct operation IDs in the results
+
+      // get just the results that apply to the current operation ID (endpoint)
+      const operationResults = testResults.filter(
+        (result) => result.operationId === operationId,
+      );
+
+      // assemble output for this endpoint
+      this.assembleEndpointLevel(operationResults, output);
```

Comment:
Since your throwing errors further down its best we don't allow a single endpoint error to prevent the rest of report from displaying (Especially if a API has 163 endpoints with only 1 such failure).  

You want to try/catch the problem here.  The endpoint with error should just include that error in the report so we know about it.

Reply:
This seems contradictory to the output format spec:

https://community.max.gov/pages/viewpage.action?pageId=2286753455

There is only a top-level error field, not one per endpoint, and there is a note indicating "no results if this value is present".

Reply:
How do you intend the logic to behave when a API has 5 endpoints and one of them throws the error in assembleEndpointLevel()?

Reply:
The two places where an Error is thrown are "we should never be here" type errors.  They should not be possible to hit unless assembleEndpointLevel is called by new code (it's private so this should be discouraged) or something changes with how OASResults.results is structured.  For the latter, this would be a change that breaks assumptions that this code relies on, specifically that there should be one and only one result object for each endpoint / example group.

It looks like I'd need to add a try/catch around where buildFromOASResult is called in order to handle when one of these errors are thrown and generate alternate json output.  Or, it could be argued that it's better to let these bubble up and let oclif handle them as a hard fault.  After all, they would not be the result of bad input or usage, rather something internal to LOAST going very wrong and resulting in it being unable to assemble proper json output.